src/data.jl

```diff
@@ -14,55 +13,60 @@ Instead of directly creating an `InferenceData`, use the exported `from_xyz` fun
 [`convert_to_inference_data`](@ref).
 """
 struct InferenceData
-    o::PyObject
-
-    function InferenceData(o::PyObject)
-        if !pyisinstance(o, arviz.InferenceData)
-            throw(ArgumentError("$o is not an `arviz.InferenceData`."))
-        end
-        return new(o)
-    end
+    groups::Dict{Symbol,Dataset}
```

Comment:
In most cases, we expect performance critical code will operate on datasets and individual arrays instead of on `InferenceData`, so in that sense backing it with a `Dict` makes sense (and is similar to how `DataFrame` is implemented). It also reduces compilation times and makes `InferenceData` mutable.

However, any function that operates on `InferenceData` will then be type-unstable, and we may need function barriers to maintain performance, so perhaps we should decide that a given `InferenceData`'s groups are immutable (as in the group cannot be swapped out, removed, or added) and use a `NamedTuple` as the backing.